lib/ace/ext/papeeria/katex-previewer.coffee

```diff
@@ -242,148 +235,69 @@ define((require, exports, module) ->
-      curToken = @tokenIterator.getCurrentToken()
-      if not curToken?
-        @outOfRange = true
-        return null
-
-      { row: tokenRow, column: tokenColumn } = @tokenIterator.getCurrentTokenPosition()
-      tokenRange = new Range(tokenRow, tokenColumn, tokenRow, tokenColumn + curToken.value.length)
-      if @range.containsRange(tokenRange)
-        @outOfRange = false
-        return curToken
-      else
-        @outOfRange = true
-        return null
-
-    stepTo: (row, column) ->
-      @tokenIterator = new TokenIterator(@session, row, column)
-      @outOfRange = not @range.contains(row, column)
-
-
   class EquationRangeHandler
-    @BEGIN_EQUATION_TOKEN_SEQUENCES: [
-      [
-        { type: "paren.rparen", value: "}" }
-        { type: "variable.parameter", value: "equation" }
-        { type: "paren.lparen", value: "{" }
-        { type: "storage.type", value: "\\begin" }
-      ]
-      [
-        { type: "paren.rparen", value: "}" }
-        { type: "variable.parameter", value: "equation*" }
-        { type: "paren.lparen", value: "{" }
-        { type: "storage.type", value: "\\begin" }
-      ]
-      [ { type: "string", value: "\\[" } ]
-      [ { type: "string", value: "\\(" } ]
-      [ { type: "string", value: "$" } ]
-      [ { type: "string", value: "$$" } ]
-    ]
-    @END_EQUATION_TOKEN_SEQUENCES: [
-      [
-        { type: "storage.type", value: "\\end" }
-        { type: "paren.lparen", value: "{" }
-        { type: "variable.parameter", value: "equation" }
-        { type: "paren.rparen", value: "}" }
-      ]
-      [
-        { type: "storage.type", value: "\\end" }
-        { type: "paren.lparen", value: "{" }
-        { type: "variable.parameter", value: "equation*" }
-        { type: "paren.rparen", value: "}" }
-      ]
-      [ { type: "string", value: "\\]" } ]
-      [ { type: "string", value: "\\)" } ]
-      [ { type: "string", value: "$" } ]
-      [ { type: "string", value: "$$" } ]
-    ]
+    @DOCUMENT_END_ERROR_CODE: "js.math_preview.error.document_end"
+    @EMPTY_LINE_ERROR_CODE: "js.math_preview.error.empty_line"
+    @WHITESPACE_LINE_ERROR_CODE: "js.math_preview.error.whitespace_line"
 
     # empty constructor
     constructor: (@editor) ->
 
-    getBoundary: (tokenIterator, start) ->
-      moveToBoundary = if start then (=> tokenIterator.stepBackward()) else (=> tokenIterator.stepForward())
-      moveFromBoundary = if start then (=> tokenIterator.stepForward()) else (=> tokenIterator.stepBackward())
-      boundarySequences = (
-        if start
-        then EquationRangeHandler.BEGIN_EQUATION_TOKEN_SEQUENCES
-        else EquationRangeHandler.END_EQUATION_TOKEN_SEQUENCES
-      )
+    getBoundary: (session, row, column, start) ->
+      summand = if start then -1 else 1
+      startRow = row
+      curIndex = session.doc.positionToIndex({ row, column }, startRow)
+      curRow = row
+      curColumn = column
+
+      while true
+        nextIndex = curIndex + summand
```

Comment:
Is index just a position in (text as string) ? Does it mean that index < 0 <=> you are out of the text start?

Reply:
This happens when we're going backwards. There is a check for that a line lower.

Reply:
that is, column?

Reply:
Yes, index from a start row. Start row is not necessarily the current row.